### ISSUE TITLE
Alerting that a clock-triggered task is delayed?

### ISSUE THREAD

**Issue by hjoliver** [cylc/cylc#2207]
It might be useful to alert users that a clock-triggered task's time has arrived, but the task cannot trigger due to other dependencies not being satisfied. 

The motivation behind this is, operational people like to know if things are running "late".  NIWA operations currently uses Nagios IT infrastructure monitoring for this sort of thing - certain tasks are expected to have reported in to Nagios by certain times.  I figure we could do it in Cylc by alerting when a clock-trigger is up but the task can't run for other reasons, because the clock-trigger represents the time that the suite writer thinks the task *should* run.

**Comment by matthewrmshin** [comment]
Do you want this to be a new task event? Or something else?

**Comment by hjoliver** [comment]
I was thinking a new task event.

**Comment by matthewrmshin** [comment]
OK. (But I am struggling to come up with a name for this event!)

**Comment by hjoliver** [comment]
Yeah, me too.  Maybe just "late" would do, since only clock-triggered tasks are connected to the wall-clock and can thus be considered to be on time or not.

**Comment by matthewrmshin** [comment]
*late* is good. I am thinking about *delayed* for a similar reason.

**Comment by oliver-sanders** [comment]
Should this event be fired whenever a clock-triggered task is running behind schedule or only when this is the result of unsatisfied up-stream dependencies?

**Comment by hjoliver** [comment]
> Should this event be fired whenever a clock-triggered task is running behind schedule or only when this is the result of unsatisfied up-stream dependencies?

I think whenever behind schedule, however, the only reason for that to occur is unsatisfied upstream dependencies isn't it?

**Comment by hjoliver** [comment]
[Note description updated]

**Comment by oliver-sanders** [comment]
> however the only reason for that to occur is unsatisfied upstream dependencies isn't it?

When starting up this example suite there will be many delayed tasks despite them having no upstream dependencies.

```
[scheduling]
    initial cycle point = 19700101T00
    [[special tasks]]
        clock-triggered = happy_new_year(PT0H)
    [[dependencies]]
        [[[0101T00]]]
            graph = happy_new_year

**Comment by hjoliver** [comment]
@oliver-sanders - you've certainly raised an interesting complication!  I guess I was only considering *existing* task proxies that can't run even though their clock-trigger times are up. In your example, each task can run as soon as it exists - they're not delayed by the clock or by dependencies - but the trouble is the suite is so far behind that large numbers of future tasks are being delayed by virtue of the fact that they don't even exist yet.  I might have to think about this...

**Comment by oliver-sanders** [comment]
This needs to wait until after #2157 has been merged.

**Comment by oliver-sanders** [comment]
Coming back to this now that we're out of rebase territory...

I think It's valid to send messages in these cases as the tasks have been "delayed". Now that messages are grouped this will only result in two emails.

**Comment by hjoliver** [comment]
@oliver-sanders - so are you suggesting this: when a task proxy with a clock trigger is instantiated it should raise a delayed/late alert if its clock-trigger time has passed, even if it can run immediately - or (as per my original suggestion) only if it is still waiting for other (task dependency) reasons?

I guess the difference is:
 1. I'm late but I can't submit!
vs
 2. I'm submitting now, but later than expected!

With 1. we'll just get one alert from the first stuck task, or no alerts if the suite simply starts way behind the clock as in your example.  With 2. we'll get an ongoing series of alerts until the suite catches up to the clock.  Another option could be a single alert from the first late instance of a task.  I'm struggling to see what's really needed here.  If your suite gets delayed, do you really want to get woken up by an alert from every successive instance of a late task until the suite catches up?

.   Maybe that's OK, but maybe users would ra

**Comment by arjclark** [comment]
> do you really want to get woken up by an alert from every successive instance of a late task until the suite catches up?

Ideally not. Perhaps it'd be possible to record the start time for the suite and ignore any clock triggered task delays for trigger times before that so doing a catch-up run wouldn't spam you? That way you wouldn't get spammed/messaged about expected delayed tasks.

**Comment by matthewrmshin** [comment]
I imagine that *1. I'm late but I can't submit!* is the main use case here?

Use case *2. I'm submitting now, but later than expected!* probably does not matter so much. If the task becomes irrelevant because it is so late, then it should probably expire.

**Comment by matthewrmshin** [comment]
Discussed. OK to ignore late warnings for tasks with a clock trigger point prior to the suite start up.

**Comment by oliver-sanders** [comment]
@dvalters Prior to #2157 I implemented this feature _(minus the part about not issuing late warnings for tasks with a clock trigger prior the the suite start time)_ in the following branch https://github.com/oliver-sanders/cylc/tree/2207.clock-trigger-delayed-event.
